tests/unit/unit3200.c

```diff
@@ -0,0 +1,152 @@
+}
+
+#ifdef __GNUC__
+#pragma GCC diagnostic ignored "-Woverlength-strings"
+#endif
+
+#define NUMTESTS 5
+static const char *filecontents[] = {
+  /* Both should be read */
+  "LINE1\n"
+  "LINE2 NEWLINE\n",
+
+  /* Both should be read */
+  "LINE1\n"
+  "LINE2 NONEWLINE",
+
+  /* Only first should be read */
+  "LINE1\n"
+  C4096,
+
+  /* First line should be read */
+  "LINE1\n"
+  C4096 "SOME EXTRA TEXT",
+
+  /* First and third line should be read */
+  "LINE1\n"
+  C4096 "SOME EXTRA TEXT\n"
+  "LINE3\n" };
+
+#ifdef __GNUC__
+#pragma GCC diagnostic warning "-Woverlength-strings"
+#endif
+
+
+UNITTEST_START
+{
+  size_t i;
+  for(i = 0; i < NUMTESTS; i++) {
+    FILE *fp;
+    char buf[4096];
+    int len = 4096;
+    char *line;
+
+    fp = fopen(TESTINPUT, "wb");
+    abort_unless(fp != NULL, "Cannot open testfile");
+    fwrite(filecontents[i], 1, strlen(filecontents[i]), fp);
+    fclose(fp);
+
+    fp = fopen(TESTINPUT, "rb");
+    abort_unless(fp != NULL, "Cannot open testfile");
+
+    switch(i) {
+      case 0:
+        line = Curl_get_line(buf, len, fp);
+        fail_unless(line && !strcmp("LINE1\n", line),
+          "First line failed (1)");
+        line = Curl_get_line(buf, len, fp);
+        fail_unless(line && !strcmp("LINE2 NEWLINE\n", line),
+          "Second line failed (1)");
+        line = Curl_get_line(buf, len, fp);
+        abort_unless(line == NULL, "Missed EOF (1)");
+        break;
+      case 1:
+        line = Curl_get_line(buf, len, fp);
+        fail_unless(line && !strcmp("LINE1\n", line),
+          "First line failed (2)");
+        line = Curl_get_line(buf, len, fp);
+        fail_unless(line && !strcmp("LINE2 NONEWLINE\n", line),
+          "Second line failed (2)");
+        line = Curl_get_line(buf, len, fp);
+        abort_unless(line == NULL, "Missed EOF (2)");
+        break;
+      case 2:
+        line = Curl_get_line(buf, len, fp);
+        fail_unless(line && !strcmp("LINE1\n", line),
+          "First line failed (3)");
+        line = Curl_get_line(buf, len, fp);
+        fail_unless(line == NULL,
+          "Did not detect max read on EOF (3)");
+        break;
+      case 3:
+        line = Curl_get_line(buf, len, fp);
+        fail_unless(line && !strcmp("LINE1\n", line),
+          "First line failed (4)");
+        line = Curl_get_line(buf, len, fp);
+        fail_unless(line == NULL,
+          "Did not ignore partial on EOF (4)");
+        break;
+      case 4:
+        line = Curl_get_line(buf, len, fp);
+        fail_unless(line && !strcmp("LINE1\n", line),
+          "First line failed (5)");
+        line = Curl_get_line(buf, len, fp);
+        fail_unless(line && !strcmp("LINE3\n", line),
+          "Third line failed (5)");
+        line = Curl_get_line(buf, len, fp);
+        abort_unless(line == NULL, "Missed EOF (5)");
+        break;
+      default:
+        abort_unless(1, "Unknown case");
+        break;
+    }
+    fclose(fp);
+  }
+  return 0;
+}
+UNITTEST_STOP
```

Comment:
I love the unit test!

Reply:
I don't have any Windows environment at hand, but I'm confused. eof() should only trigger on the condition of having reached the end of input, not some control character. So my guess would be that the curl_get_line() function should just return the string including whatever control character is in there. It is then up to the parser that calls curl_get_line to decide what to do with that.

I added a line containing ^Z in the unit test just to check what Windows would make of it. 

Maybe the code needs something like `#ifndef WIN32` to not break compatibility in WIndows - which was the whole purpose of the PR anyway - maximum downwards compatibility.

Reply:
According to the OK result of the changed unit test on Windows, curl_get_line() seems to correctly return a string including the ^Z character. So I guess the fail in test 8 is caused by differences in parsing the cookie on Windows compared to Linux or MacOS.

What to do? Remove the offending line from test 8 won't fix the difference in test output.

Reply:
Hold on, the test was not quite conclusive, retrying, waiting for Windows output.

Reply:
I think the difference in processing is caused by cookie.c:1267 opening the file as FOPEN_READTEXT,  which makes Window interpret (some) control characters. The previous version of curl_get_line() would discard the line as a partial read because of the missing \n. The new version reads the data up until, but excluding the ^Z, so the cookie.c:452 invalid_octets() function does not throw the data away. What's left is a valid `cookie26=` cookie. 

Still no clue how to handle this, I guess there's good reasons to use the FOPEN_READTEXT.

Reply:
As I recall things we went with text file simply because cookies are text. there were no considerations made for silly ^Z-handling. I just submitted #10017 to change this, let's see how that goes :smile: